sw/device/silicon_creator/rom/doc/rom_customization.md

```diff
@@ -0,0 +1,98 @@
+---
+Document: RFC
+Title: ROM Vendor Customizations
+Status: Draft
+Authors: Samuel Ortiz <sameo@opentitan.org>
+Reviewers:
+---
+
+# RFC: ROM Vendor Customizations 
+
+## Problem Statement
+
+Currently, OpenTitan mask ROM implementations lack customization options for silicon creators and vendors. While discrete OpenTitan ROMs can be directly built upon the [upstream implementation](https://github.com/lowRISC/opentitan/tree/master/sw/device/silicon_creator/rom), the same isn't true for integrated OpenTitan top ROMs.
+
+Integrated OpenTitan ROMs configure, initialize, and interact with external subsystems specific to the platform. These interactions often vary between vendors and platform variants. A single, fixed ROM implementation prevents downstream silicon creators and vendors from directly utilizing the upstream code, forcing them to fork the project and maintain custom versions internally.
+
+## Requirements
+
+To enable direct vendor contribution and consumption of OpenTitan upstream ROM implementations, the project architecture and source code for ROMs must meet the following requirements:
+
+1. Vendor customization: OpenTitan ROM code should allow silicon creators and vendors to customize the upstream ROM execution flow with vendor-specific initialization and configuration sequences.
+2. Controlled customization points: OpenTitan ROMs should be customizable at specific, predefined execution points. Vendors can only modify ROM behavior at these designated points.
+3. Support for closed-source customization: The OpenTitan ROM build infrastructure should support linking ROM images with externally stored, vendor-specific customization code, allowing vendors to keep their customizations private.
+4. Attestation: Silicon creators and vendors should provide endorsed reference values for their customized OpenTitan ROMs, including the ROM version and its associated hash. Out-of-band verifiers can then compare these values with the ROM's generated attestation reports.
+5. Default functionality: By default, OpenTitan ROMs should function without any vendor-specific customizations.
+
+## Proposed Approach
+
+At a high level, the proposed solution involves implementing an OpenTitan ROM as a state machine. Each state in this state machine is semantically defined by the upstream ROM code and represents a customizable execution step.
```

Comment:
Comment from @modix in https://github.com/lowRISC/opentitan/issues/25633#issuecomment-2549678655:

State machine considerations:

We are still working on evaluating the effectiveness of hardware countermeasures with respect to fault attacks. Until then, we recommend to maintain the level of hardening present in the upstream version of the ROM.

In practice, it is very difficult to harden code with too many branches. Ideally, we should aim as much as possible to maintain a linear execution sequence. I propose we define a requirement to cover this.

Reply:
> We are still working on evaluating the effectiveness of hardware countermeasures with respect to fault attacks. Until then, we recommend to maintain the level of hardening present in the upstream version of the ROM.
> 

One interesting side effect of defining the ROM as a state machine is that you can systematically include CFI within the state machine walking routine. I have a pending patch to show this in practice, but the idea is that each ROM state would be protected by its implicit CFI counter so that the ROM state machine flow is controlled.

> In practice, it is very difficult to harden code with too many branches. Ideally, we should aim as much as possible to maintain a linear execution sequence. I propose we define a requirement to cover this.

That makes sense, yes. To be clear here, the current ROM is not linear and has many branches, mostly for error handling but also for the bootstrap and immutable ROM extension flows for example. If anything, I feel that explicitly defining the ROM as a state machine would allow for more easily tracking those branches than the current implementation.

Reply:
> > We are still working on evaluating the effectiveness of hardware countermeasures with respect to fault attacks. Until then, we recommend to maintain the level of hardening present in the upstream version of the ROM.
> 
> One interesting side effect of defining the ROM as a state machine is that you can systematically include CFI within the state machine walking routine. I have a pending patch to show this in practice, but the idea is that each ROM state would be protected by its implicit CFI counter so that the ROM state machine flow is controlled.
> 

I added a couple of commits to this PR to illustrate how CFI counters could be inserted in the ROM state machine walker.